### PR TITLE
New resource view: Network -> Ingress Classes

### LINKED ISSUE #2576: Add a view of  kind: IngressClass
**What would you like to be added**:
Add a visibility for kind: IngressClass

**Why is this needed**:
Right now we have no UI representation of kind: IngressClass

**Environment you are Lens application on:**

- Kubernetes distribution: EKS 1.18 | EKS 1.19
- Desktop OS: Ubuntu 20.04

**API DOCS:**

- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#ingressclass-v1-networking-k8s-io
- https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class

### PR DESCRIPTION
**Pull request by ixrock**
https://user-images.githubusercontent.com/6377066/210792819-7bad39d9-307c-4328-8e4d-ddfd459bed99.mov

_TODO:_

- [x] implement "Set as default" from ingress class context menu item
- [x] fix: [duplicating items](https://github.com/lensapp/lens/pull/6808#issuecomment-1366688120) on every page visiting _(ingress class listing)_
- [x] fix: [editing](https://github.com/lensapp/lens/pull/6808#issuecomment-1367909330) items doesn't work
- [x] inject more fields to Drawer details panel (e.g. from `spec.parameters.*`)

close https://github.com/lensapp/lens/issues/2576

### CODE REVIEW

**Review by aleksfront — COMMENTED** [review]
Looks good! Please fix the conflicts.

**Comment by aleksfront** [comment]
While testing, created 2 simple `IngressClasses` from the [docs](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class).

* After some time, 2 additional entries appeared with `IngressParameter` type. Is this how it should be?
* 2 or more items gets selected at once
* Created `IngressClass` with `controller` parameter, but `Controller` column always empty.

All issues above demonstrated in this video:


https://user-images.githubusercontent.com/9607060/209669124-0bc56e62-c945-4d94-a1bf-0f764942130b.mov

**Comment by ixrock** [comment]
Some another weirdness actually with repeating initial values in store on view refresh? _Keeps digging.._

https://user-images.githubusercontent.com/6377066/209825908-9b8d5ea3-9ed6-4d67-b677-c10d78c1187b.mov

File: src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+import { getInjectable, lifecycleEnum } from "@ogre-tools/injectable";
+import type { IngressClass } from "../../../common/k8s-api/endpoints/ingress-class.api";
+
+// TODO: figure out if this needs to be injectable
+export const ingressClassSetDefaultInjectable = getInjectable({
```
Comment (jansav) [inline src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts:11]: This is not used anywhere, so remove?

File: src/renderer/components/+network-ingresses/ingresses-sidebar-items.injectable.tsx [hunk 2]
```diff
@@ -3,30 +3,48 @@
...
+        orderNumber: 30,
+      },
+      {
+        id: "ingressclasses",
+        parentId: networkSidebarItemId,
+        title: "Ingress Classes",
+        onClick: di.inject(navigateToIngressClassesInjectable),
+        isActive: di.inject(routeIsActiveInjectable, ingressClassRoute),
+        isVisible: ingressClassRoute.isEnabled,
         orderNumber: 30,
```
Comment (jansav) [inline src/renderer/components/+network-ingresses/ingresses-sidebar-items.injectable.tsx]: Different order number should be used to make sure that the ordering is consistent.

File: src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts (hunk 1 above)
Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts:11]: It was preparation.. Please check again.

File: src/renderer/components/+network-ingresses/ingresses-sidebar-items.injectable.tsx (hunk 2 above)
Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingresses-sidebar-items.injectable.tsx]: Fixed.

**Comment by ixrock** [comment]
One more issue found, editing `IngressClass` doesn't work from item's menu:

Screenshot 2022-12-30 at 15 05 02

**UPD:** fixed with `static namespaced = false` for `IngressClass` api object.

File: src/common/front-end-routing/routes/cluster/network/ingress-class/ingress-classeses-route.injectable.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import {
+  shouldShowResourceInjectionToken,
+} from "../../../../../cluster-store/allowed-resources-injection-token";
+
+const ingressClassesesRouteInjectable = getInjectable({
+  id: "ingress-classes-route",
+
+  instantiate: (di) => {
+    const isEnabled = di.inject(shouldShowResourceInjectionToken, {
+      apiName: "ingresses",
```
Comment (Nokel81) [inline src/common/front-end-routing/routes/cluster/network/ingress-class/ingress-classeses-route.injectable.ts]: shouldn't this be `ingressclasses`?

Comment (ixrock) [reply src/common/front-end-routing/routes/cluster/network/ingress-class/ingress-classeses-route.injectable.ts]: yeah, most probably.. good catch!

File: src/common/k8s-api/endpoints/ingress-class.api.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    super({
+      objectConstructor: IngressClass,
+      checkPreferredVersion: true,
+      fallbackApiBases: ["/apis/extensions/v1beta1/ingressclasses"],
+    });
+  }
+}
+
+// API docs: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#ingressclass-v1-networking-k8s-io
+export type IngressClassMetadata = KubeObjectMetadata<KubeObjectScope.Namespace> & {
```
Comment (Nokel81) [inline src/common/k8s-api/endpoints/ingress-class.api.ts]: This should be `KubeObjectScope.Cluster` too

Comment (ixrock) [reply src/common/k8s-api/endpoints/ingress-class.api.ts]: Fixed

**Comment by ixrock** [comment]
@lensapp/lens-maintainers PTAL again.

File: src/common/k8s-api/endpoints/ingress-class.api.ts [hunk 5]
```diff
@@ -16,17 +17,17 @@ export class IngressClassApi extends KubeApi<IngressClass> {
...
             [IngressClass.ANNOTATION_IS_DEFAULT]: JSON.stringify(isDefault),
-          }
-        }
-      }
+          },
+        },
+      },
     }, {
       headers: {
         "content-type": "application/strategic-merge-patch+json",
```
Comment (Nokel81) [inline src/common/k8s-api/endpoints/ingress-class.api.ts:33]: nit: remove comment.

File: src/common/k8s-api/endpoints/ingress-class.api.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  getCtrlApiGroup() {
+    return this.spec?.parameters?.apiGroup;
+  }
+
+  getCtrlScope() {
+    return this.spec?.parameters?.scope;
+  }
+
+  getCtrlNs() {
+    return this.spec?.parameters?.namespace as string;
```
Comment (Nokel81) [inline src/common/k8s-api/endpoints/ingress-class.api.ts]: why `as string`?

File: src/common/rbac.ts [hunk 7]
```diff
@@ -87,6 +87,11 @@ export const apiResourceRecord: Record<KubeResource, KubeApiResourceData> = {
     group: "networking.k8s.io",
     namespaced: true,
   },
+  ingressclasses: {
+    kind: "IngressClass",
+    group: "networking.k8s.io",
+    namespaced: true,
```
Comment (Nokel81) [inline src/common/rbac.ts]: As we found out before, not `namespaced`.

File: src/renderer/components/+network-ingresses/ingress-class-menu.injectable.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  kubeObjectMenuItemInjectionToken,
+} from "../kube-object-menu/kube-object-menu-item-injection-token";
+import { ingressClassSetDefaultInjectable } from "./ingress-class-set-default.injectable";
+import { MenuItem } from "../menu";
+import type { IngressClass } from "../../../common/k8s-api/endpoints/ingress-class.api";
+import type { KubeObjectMenuProps } from "../kube-object-menu";
+import { Icon } from "../icon";
+
+export interface IngressClassMenuProps extends KubeObjectMenuProps<IngressClass> {
+  setDefault(item: IngressClass): void;
```
Comment (Nokel81) [inline src/renderer/components/+network-ingresses/ingress-class-menu.injectable.tsx]: nit: this should be a dependency and not a `Prop`. That way we shouldn't have to have the `as KubeObjectMenuItemComponent` below.

File: src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      const api = di.inject(ingressClassApiInjectable);
+      const store = di.inject(ingressClassStoreInjectable);
+
+      const defaultIngressClassesUpdate = store.items
+        .filter((item: IngressClass) => item.isDefault && currentItem !== item)
+        .map(item => api.setAsDefault({ name: item.getName() }, false));
+
+      await Promise.all(defaultIngressClassesUpdate);
+      await api.setAsDefault({ name: currentItem.getName() });
+      await store.reloadAll({ force: true });
```
Comment (Nokel81) [inline src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts]: Why is the store needed to be reloaded, shouldn't the `subscribe` just get the updates normally?

File: src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import ingressClassApiInjectable
+  from "../../../common/k8s-api/endpoints/ingress-class.api.injectable";
+import ingressClassStoreInjectable from "./ingress-class-store.injectable";
+
+export const ingressClassSetDefaultInjectable = getInjectable({
+  id: "ingressClassSetDefaultInjectable",
+
+  instantiate(di) {
+    return async (currentItem: IngressClass) => {
+      const api = di.inject(ingressClassApiInjectable);
```
Comment (Nokel81) [inline src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts]: nit: dependencies should be `injected` outside of the returned implementation.

That way the timing of being ready is guaranteed.

File: src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts (hunk 9 above)
Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts:25]: Probably yes, I'll double check.

Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts]: Yep, it works, but not immediately like with `reloadAll()`, removed.

File: src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts (hunk 10 above)
Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingress-class-set-default.injectable.ts]: Fixed

File: src/common/rbac.ts (hunk 7 above)
Comment (ixrock) [reply src/common/rbac.ts]: Right, fixed.

File: src/common/k8s-api/endpoints/ingress-class.api.ts (hunk 6 above)
Comment (ixrock) [reply src/common/k8s-api/endpoints/ingress-class.api.ts]: Not sure where it comes from.. removed.

File: src/renderer/components/+network-ingresses/ingress-class-menu.injectable.tsx (hunk 8 above)
Comment (ixrock) [reply src/renderer/components/+network-ingresses/ingress-class-menu.injectable.tsx]: Not sure if we can get rid of `as KubeObjectMenuItemComponent`.
I've extracted this prop to separated interface `Dependencies` and I still see the errors:

Screenshot 2023-01-05 at 18 00 49

Other usages of `as KubeObjectMenuItemComponent`:

image

Comment (Nokel81) [reply src/renderer/components/+network-ingresses/ingress-class-menu.injectable.tsx]: Oh okay. I guess that makes sense from a typing perspective. Though it should be listed as a dep anyway.

**Review by Nokel81 — APPROVED** [review]
🎉 LGTM

**Pull request merged**